craft_application/services/init.py

```diff
@@ -27,12 +30,47 @@
 from craft_application.errors import InitError
 from craft_application.git import GitError, GitRepo, is_repo, parse_describe
 
+from ..models.constraints import MESSAGE_INVALID_NAME, PROJECT_NAME_COMPILED_REGEX
 from . import base
 
+if typing.TYPE_CHECKING:  # pragma: no cover
+    from craft_application.application import AppMetadata
+    from craft_application.services import ServiceFactory
+
 
 class InitService(base.AppService):
     """Service class for initializing a project."""
 
+    def __init__(
+        self,
+        app: AppMetadata,
+        services: ServiceFactory,
+        *,
+        default_name: str = "my-project",
+        name_regex: Pattern[str] = PROJECT_NAME_COMPILED_REGEX,
+        invalid_name_message: str = MESSAGE_INVALID_NAME,
+    ) -> None:
+        super().__init__(app, services)
+        self._default_name = default_name
+        self._name_regex = name_regex
+        self._invalid_name_message = invalid_name_message
+
+    def validate_project_name(self, name: str, *, use_default: bool = False) -> str:
+        """Validate that ``name`` is valid as a project name.
+
+        If ``name`` is invalid and ``use_default`` is False, then an InitError
+        is raised. If ``use_default`` is True, the default project name provided
+        to the service's constructor is returned.
+
+        If ``name`` is valid, it is returned.
+        """
+        if not self._name_regex.match(name):
```

Comment:
Just as a side note: if you'll use regex to validate that, it won't check whether your name has an appropriate length.

Reply:
So I thought about that but it turns out the regex itself (both here and in rockcraft) already ensures a minimum length - for instance, craft-application's regex needs at least one [a-z] so the string will always have at least one character.
I added some more valid/invalid test cases though, thanks for pointing this out.